File: java/sodium/src/sodium/time/TimerSystem.java
package sodium.time;

import sodium.*;
import java.util.LinkedList;
import java.util.Optional;

public class TimerSystem<T extends Comparable> {
    public TimerSystem(TimerSystemImpl<T> impl) {
        this.impl = impl;
        CellSink<T> timeSnk = new CellSink<T>(impl.now());
        time = timeSnk;
        Transaction.onStart(() -> {
            T t = impl.now();
            impl.runTimersTo(t);
            while (true) {
                Event ev;
                // Pop all events earlier than t.
                synchronized (eventQueue) {
                    ev = eventQueue.peekFirst();
                    if (ev != null && ev.t.compareTo(t) <= 0)
                        eventQueue.removeFirst();
                    else
                        ev = null;
                }
                if (ev != null) {
                    timeSnk.send(ev.t);
                    ev.sAlarm.send(ev.t);
                }
                else
                    break;
            }
            timeSnk.send(t);
        });
    }

    private final TimerSystemImpl<T> impl;
    public final Cell<T> time;

    private class Event {
        Event(T t, StreamSink<T> sAlarm) {
            this.t = t;
            this.sAlarm = sAlarm;
        }
        T t;
        StreamSink<T> sAlarm;
    };
    private LinkedList<Event> eventQueue = new LinkedList<>();

    private static class CurrentTimer {
        Optional<Timer> oTimer = Optional.empty();
    };

    /**
     * A timer that fires at the specified time.
     */
    public Stream<T> at(Cell<Optional<T>> tAlarm) {
        StreamSink<T> sAlarm = new StreamSink<>();
        CurrentTimer current = new CurrentTimer();
        Listener l = tAlarm.listen(oAlarm -> {
            if (current.oTimer.isPresent())
                current.oTimer.get().cancel();
            current.oTimer = oAlarm.isPresent()
                ? Optional.<Timer>of(
                    impl.setTimer(oAlarm.get(), () -> {
                        synchronized (eventQueue) {
                            eventQueue.add(new Event(oAlarm.get(), sAlarm));
                        }
                        // Open and close a transaction to trigger queued
                        // events to run.
                        Transaction.runVoid(() -> {});
                    }))
                : Optional.<Timer>empty();
        });
        return sAlarm.addCleanup(l);
    }
}

File: dotnet/src/Sodium/CSharp/Sodium.Frp/Time/TimerSystem.cs
using System;
using System.Collections.Generic;
using Sodium.Functional;

namespace Sodium.Frp.Time
{
    public class TimerSystem<T> : ITimerSystem<T>
        where T : IComparable<T>
    {
        private readonly ITimerSystemImplementation<T> implementation;

        private readonly Queue<Event> eventQueue = new Queue<Event>();

        public TimerSystem(ITimerSystemImplementation<T> implementation, Action<Exception> handleException)
        {
            this.implementation = implementation;
            this.implementation.Start(handleException);
            BehaviorSink<T> timeSink = new BehaviorSink<T>(this.implementation.Now);
            this.Time = timeSink;
            Transaction.OnStart(
                () =>
                {
                    T t = this.implementation.Now;
                    this.implementation.RunTimersTo(t);
                    while (true)
                    {
                        Event ev = null;
                        // Pop all events earlier than t.
                        lock (this.eventQueue)
                        {
                            if (this.eventQueue.Count > 0)
                            {
                                Event tempEvent = this.eventQueue.Peek();
                                if (tempEvent != null && tempEvent.Time.CompareTo(t) <= 0)
                                {
                                    ev = this.eventQueue.Dequeue();
                                }
                            }
                        }

                        if (ev != null)
                        {
                            timeSink.Send(ev.Time);
                            ev.Alarm.Send(ev.Time);
                        }
                        else
                        {
                            break;
                        }
                    }

                    timeSink.Send(t);
                });
        }

        /// <summary>
        ///     Gets a behavior giving the current clock time.
        /// </summary>
        public Behavior<T> Time { get; }

        private class Event
        {
            internal Event(T time, StreamSink<T> alarm)
            {
                this.Time = time;
                this.Alarm = alarm;
            }

            internal readonly T Time;
            internal readonly StreamSink<T> Alarm;
        }

        /// <summary>
        ///     A timer that fires at the specified time.
        /// </summary>
        /// <param name="t">The time to fire at.</param>
        /// <returns>A stream which fires at the specified time.</returns>
        public Stream<T> At(Cell<Maybe<T>> t)
        {
            StreamSink<T> alarm = new StreamSink<T>();
            Maybe<ITimer> currentTimer = Maybe.None;
            IListener l = t.Listen(
                m =>
                {
                    currentTimer.MatchSome(timer => timer.Cancel());
                    currentTimer = m.Match(
                        time => Maybe.Some(
                            this.implementation.SetTimer(
                                time,
                                () =>
                                {
                                    lock (this.eventQueue)
                                    {
                                        this.eventQueue.Enqueue(new Event(time, alarm));
                                    }
                                    // Open and close a transaction to trigger queued
                                    // events to run.
                                    Transaction.RunVoid(() => { });
                                })),
                        () => Maybe.None);
                });
            return alarm.AttachListener(l);
        }
    }
}